packages/tokenbank_v1/contracts/modules/security/RecoveryModule.sol

```diff
@@ -0,0 +1,155 @@
+        uint _lockPeriod
+        )
+        public
+        SecurityModule(_securityStorage)
+    {
+        // TODO(which should be large?)
+        recoveryPeriod = _recoveryPeriod;
+        lockPeriod = _lockPeriod;
+    }
+
+    function startRecovery(
+        address   wallet,
+        address[] calldata signers,
+        address   newOwner
+        )
+        external
+        nonReentrant
+        onlyFromMetaTx
+        notWalletOwner(wallet, newOwner)
+    {
+        require(newOwner != address(0), "ZERO_ADDRESS");
+
+        WalletRecovery storage recovery = wallets[wallet];
+        require(recovery.completeAfter == 0, "ALREAY_STARTED");
+
+        uint guardianCount = securityStorage.numGuardians(wallet);
+        require(signers.length >= (guardianCount + 1)/2, "NOT_ENOUGH_SIGNER");
+        require(isWalletOwnerOrGuardian(wallet, signers), "UNAUTHORIZED");
+
+        recovery.newOwner = newOwner;
+        recovery.completeAfter = now + recoveryPeriod;
+        recovery.guardianCount = guardianCount;
+
+        securityStorage.setLock(wallet, now + lockPeriod);
+
+        emit RecoveryStarted(wallet, newOwner, recovery.completeAfter);
+    }
+
+    function cancelRecovery(
+        address   wallet,
+        address[] calldata signers
+        )
+        external
+        nonReentrant
+        onlyFromMetaTx
+    {
+        WalletRecovery storage recovery = wallets[wallet];
+        require(recovery.completeAfter > 0, "NOT_STARTED");
+
+        uint guardianCount = wallets[wallet].guardianCount;
+        require(signers.length >= (guardianCount + 1) / 2, "NOT_ENOUGH_SIGNER");
+        require(isWalletOwnerOrGuardian(wallet, signers), "UNAUTHORIZED");
+
+        delete wallets[wallet];
+        securityStorage.setLock(wallet, 0);
+
+        emit RecoveryCancelled(wallet);
+    }
+
+    function completeRecovery(address wallet)
+        external
+        nonReentrant
+        onlyFromMetaTx
+    {
+        WalletRecovery storage recovery = wallets[wallet];
+        require(recovery.completeAfter > 0, "NOT_STARTED");
+        require(recovery.completeAfter < now, "TWO_EARLY");
+
+        Wallet(wallet).setOwner(recovery.newOwner);
+
+        delete wallets[wallet];
+        securityStorage.setLock(wallet, 0);
+
+        emit RecoveryCompleted(wallet, recovery.newOwner);
+    }
+
+    function extractMetaTxSigners(
+        address /*wallet*/,
+        bytes4  method,
+        bytes   memory data
+        )
+        internal
+        view
+        returns (address[] memory signers)
+    {
+        require(
+            method == this.startRecovery.selector ||
+            method == this.completeRecovery.selector ||
+            method == this.cancelRecovery.selector,
+            "INVALID_METHOD"
+        );
+
+        // data layout: {length:32}{sig:4}{_wallet:32}{signer1:32}{signer2:32}{...}
+        require((data.length - 68) % 32 == 0, "INVALID_DATA");
+        uint numSigners = (data.length - 68) / 32;
```

Comment:
I think this should be `(data.length - 36)` because the 32 bytes length value isn't part of `data.length`.

Reply:
// data layout: {length:32}{sig:4}{_wallet:32}{signer1:32}{signer2:32}{...}

The first signer address is preceded with 32 bytes length, 4 bytes method selector, and 32 bytes wallet address, so it has 68 bytes ahead of it.

Reply:
But I think there is a bug, I'll fix it and please review this part again.